src/key.rs

```diff
@@ -45,61 +45,272 @@ use crate::{
+    /// physical access applications, such as providing physical access to buildings via
+    /// PIV-enabled door locks. The end user PIN is NOT required to perform private key
+    /// operations for this slot.
+    CardAuthentication,
+
+    /// These slots are only available on the YubiKey 4 & 5. They are meant for previously
+    /// used Key Management keys to be able to decrypt earlier encrypted documents or
+    /// emails. In the YubiKey 4 & 5 all 20 of them are fully available for use.
+    Retired(RetiredSlotId),
+
+    /// This slot is only available on YubiKey version 4.3 and newer. It is only used for
+    /// attestation of other keys generated on device with instruction `f9`. This slot is
+    /// not cleared on reset, but can be overwritten.
+    Attestation,
+}
+
+impl TryFrom<u8> for SlotId {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            YKPIV_KEY_AUTHENTICATION => Ok(SlotId::Authentication),
+            YKPIV_KEY_SIGNATURE => Ok(SlotId::Signature),
+            YKPIV_KEY_KEYMGM => Ok(SlotId::KeyManagement),
+            YKPIV_KEY_CARDAUTH => Ok(SlotId::CardAuthentication),
+            YKPIV_KEY_ATTESTATION => Ok(SlotId::Attestation),
+            _ => RetiredSlotId::try_from(value).map(SlotId::Retired),
+        }
+    }
+}
+
+impl From<SlotId> for u8 {
+    fn from(slot: SlotId) -> u8 {
+        match slot {
+            SlotId::Authentication => YKPIV_KEY_AUTHENTICATION,
+            SlotId::Signature => YKPIV_KEY_SIGNATURE,
+            SlotId::KeyManagement => YKPIV_KEY_KEYMGM,
+            SlotId::CardAuthentication => YKPIV_KEY_CARDAUTH,
+            SlotId::Retired(retired) => retired.into(),
+            SlotId::Attestation => YKPIV_KEY_ATTESTATION,
         }
-        _ => return Err(Error::InvalidObject),
-    };
+    }
+}
+
+impl SlotId {
+    /// Returns the [`ObjectId`] that corresponds to a given [`SlotId`].
+    pub(crate) fn object_id(self) -> ObjectId {
+        match self {
+            SlotId::Authentication => YKPIV_OBJ_AUTHENTICATION,
+            SlotId::Signature => YKPIV_OBJ_SIGNATURE,
+            SlotId::KeyManagement => YKPIV_OBJ_KEY_MANAGEMENT,
+            SlotId::CardAuthentication => YKPIV_OBJ_CARD_AUTH,
+            SlotId::Retired(retired) => retired.object_id(),
+            SlotId::Attestation => YKPIV_OBJ_ATTESTATION,
+        }
+    }
+}
 
-    Ok(id)
+/// Retired slot IDs.
+#[allow(missing_docs)]
+#[derive(Clone, Copy, Debug)]
+pub enum RetiredSlotId {
+    R1,
+    R2,
+    R3,
+    R4,
+    R5,
+    R6,
+    R7,
+    R8,
+    R9,
+    R10,
+    R11,
+    R12,
+    R13,
+    R14,
+    R15,
+    R16,
+    R17,
+    R18,
+    R19,
+    R20,
+}
+
+impl TryFrom<u8> for RetiredSlotId {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            YKPIV_KEY_RETIRED1 => Ok(RetiredSlotId::R1),
```

Comment:
In other places (e.g. `StatusWords`, and throughout the `yubihsm` crate) I've eliminated these constants in favor of using the literal codes for these conversions between the `u8` representation and the enum variants.

I guess it helps statically assert they're the same for the same labels, since you have to write them twice.